### ISSUE TITLE
JuMPContainer iteration thoughts

### ISSUE THREAD

**Issue by IainNZ** [JuliaOpt/JuMP.jl#348]
```
m = Model()
nums = [1.0, 2.0, 3.0, 4.0, 5.0]
letters = ["A", "B", "C", "D"]
@defVar(m, Flow[nums,letters])
for (index, value) in Flow
  @show index
  @show value
end
```

Gives output like

```
index => 4.0
value => "C"
```

which confused someone, and then confused me too.

**Comment by IainNZ** [comment]
Ah, ok, I get what its doing. Its iterating over just the keys.

**Comment by IainNZ** [comment]
Actually it isn't. I'm getting things mixed up. 

Consider:

```julia
A = rand(2,2)
for iter in A
  @show iter
end

A = [("A",2) => nothing, ("B",3) => nothing]
for iter in A
  @show iter
end

using JuMP
m = Model()
@defVar(m, ja[1:2,1:2], Bin)
@defVar(m, jd[["A","B"],[2,3]], Bin)
solve(m)
ja = getValue(ja)
for iterja in ja
  @show iterja
end
jd = getValue(jd)
for iterjd in jd
  @show iterjd
end
```

which has output

```
iter => 0.520622776314158
iter => 0.7107648975082581
iter => 0.3833170741316041
iter => 0.019295811653074146
iter => (("A",2),nothing)
iter => (("B",3),nothing)
iterja => (1,1,0.0)
iterja => (2,1,0.0)
iterja => (1,2,0.0)
iterja => (2,2,0.0)
iterjd => ("B",2,0.0)
iterjd => ("A",2,0.0)
iterjd => ("A",3,0.0)
iterjd => ("B",3,0.0)
```

Maybe we should be returning something more like a Julia dict for JuMPContainers, which would let you do `(index, value)` sorta. The problem is the `index` couldn't be used as-is as an index, you'd need to splat it `index...`

**Comment by joehuchette** [comment]
The problem is that ``JuMPContainer``s have indexing kind of like arrays and kind of like dictionaries. Since you (should) know at all times how many indices a container has, you could always write, e.g.
```julia
for (i,j,k,v) in x
```
for a _given_ v. This makes writing code for a generic ``JuMPContainer`` difficult, and the printing code has lots of things like
```julia
for it in x
    v = x[end]
    ...
end
```
Not sure if this is worse than splatting the indices at all time.

**Comment by joehuchette** [comment]
Resolved with the deprecation in #516
